Add esp32p4 support

Vincent1-python opened the pull request:
<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->

Add esp32p4 support.
### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->
All completed after https://github.com/micropython/micropython/pull/17356, except for DAC not supported by hardware.

### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

     Delete this heading if not relevant (i.e. small fixes) -->

Vincent1-python committed:
esp32/boards/ESP32_GENERIC_P4: Add board definition for ESP32P4.

Includes a base variant with LAN, and C5_WIFI and C6_WIFI variants with
LAN, WiFi and BLE.

And builds this board in the esp32 CI, to cover the P4 support.

Signed-off-by: Vincent1-python <pywei201209@163.com>
Signed-off-by: Angus Gratton <angus@redyak.com.au>
Signed-off-by: Damien George <damien@micropython.org>

Vincent1-python committed:
esp32: Add support for ESP32-P4.

This commit adds support for ESP32-P4 SoCs.

Signed-off-by: Vincent1-python <pywei201209@163.com>
Signed-off-by: Angus Gratton <angus@redyak.com.au>
Signed-off-by: Damien George <damien@micropython.org>

projectgus commented on ports/esp32/machine_sdcard.c:
```diff
@@ -406,9 +410,20 @@ static mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
-
-        DEBUG_printf("  Calling init_slot()");
-        check_esp_err(sdmmc_host_init_slot(self->host.slot, &slot_config));
+        #if MP_SD_LDO
```
This macro will need a clearer name, something like `#define MICROPY_BOARD_SDCARD_LDO_CH 4` and then can test `#ifdef MICROPY_BOARD_LDO_CH`.

However as this is only needed for the FireBeetle then suggest moving it to a second PR.

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -29,7 +29,7 @@
 #include "driver/gpio.h"
 
-#if SOC_TOUCH_SENSOR_SUPPORTED
+#if SOC_TOUCH_SENSOR_SUPPORTED //&& !CONFIG_IDF_TARGET_ESP32P4
```
```suggestion
#if SOC_TOUCH_SENSOR_SUPPORTED
```

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -43,16 +43,27 @@
 #include "driver/touch_pad.h"
 #elif SOC_TOUCH_SENSOR_VERSION == 2 // All other SoCs with touch, to date
```
```suggestion
#elif SOC_TOUCH_SENSOR_VERSION == 2 // Most ESP32 SoCs
```

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -43,16 +43,27 @@
 #include "driver/touch_pad.h"
 #elif SOC_TOUCH_SENSOR_VERSION == 2 // All other SoCs with touch, to date
 #include "driver/touch_sensor.h"
+#elif SOC_TOUCH_SENSOR_VERSION == 3 // All other SoCs with touch, to date
```
```suggestion
#elif SOC_TOUCH_SENSOR_VERSION == 3 // Currently ESP32-P4

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -43,16 +43,27 @@
+static touch_sensor_handle_t  s_sens_handle = NULL;
+static touch_channel_handle_t s_chan_handle[15] = {};
+uint32_t benchmark[14] = {};
+int i;
```
These two variables look like they can be local variables inside functions, not globals.

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -102,14 +128,53 @@ static mp_obj_t mtp_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_
+	touch_sensor_config_t sens_cfg = TOUCH_SENSOR_DEFAULT_BASIC_CONFIG(3, sample_cfg);
+        esp_err_t err = touch_sensor_new_controller(&sens_cfg, &s_sens_handle);
+	if (err != ESP_OK) {
+    	   mp_raise_ValueError(MP_ERROR_TEXT("Touch pad error 1"));
```
I can see why you've numbered these (`1`, `2`, etc) but it's not something that we do in MicroPython.

Suggest either making the errors the same or add some better description of what failed. Most of these errors should never happen unless there's an internal driver bug, yes?

projectgus commented on ports/esp32/machine_touchpad.c:
```diff
@@ -128,11 +193,13 @@ static mp_obj_t mtp_config(mp_obj_t self_in, mp_obj_t value_in) {
     #elif SOC_TOUCH_SENSOR_VERSION == 2
     esp_err_t err = touch_pad_config(self->touchpad_id);
+    #elif SOC_TOUCH_SENSOR_VERSION == 3
+    esp_err_t err = ESP_OK;
```
If this function doesn't do anything then we should `#ifdef` it out and not include it on the object for version 3

projectgus commented on ports/esp32/modmachine.c:
```diff
@@ -233,8 +237,17 @@ MP_NORETURN static void machine_bootloader_rtc(void) {
+    esp_rom_gpio_pad_select_gpio(GPIO_NUM_35);
+    gpio_set_direction(GPIO_NUM_35, GPIO_MODE_INPUT);
+    gpio_set_pull_mode(GPIO_NUM_35, GPIO_PULLDOWN_ONLY);
+    if (gpio_get_level(GPIO_NUM_35) == 0) {
```
This seems very board-specific ... ? Is there really no way to force a download boot reset?

Vincent1-python replied:
I haven't found anything that can be realized with the register value of RTC at present, so I'll use this first.

Vincent1-python replied:
But it can also be used normally.

projectgus commented on ports/esp32/network_wlan.c:
```diff
@@ -767,7 +767,7 @@ static const mp_rom_map_elem_t wlan_if_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_SEC_WPA2_WPA3_ENT), MP_ROM_INT(WIFI_AUTH_WPA2_WPA3_ENTERPRISE) },
     #endif
-    #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 5, 0)
+    #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 5, 0) && !CONFIG_IDF_TARGET_ESP32P4
```
What's the root cause of needing this exclusion?

Vincent1-python replied:
I think I should.

Vincent1-python replied:
Maybe the versions of my two components are too low. I'll try to update them to solve the problem.

projectgus commented on ports/esp32/mpconfigport.h:
```diff
@@ -274,15 +276,25 @@
 #define MICROPY_HW_USB_CDC                  (MICROPY_HW_ENABLE_USBDEV)
 #endif
 
+// Enable stdio over native USB peripheral HID via TinyUSB
```
This feature is unrelated to this PR, but also it's probably not a feature we will add to MicroPython: we already have the [machine.USBDevice](https://docs.micropython.org/en/latest/library/machine.USBDevice.html) interface which allows creating any custom USB HID device with Python code, instead.

(Suggest removing it from this PR, at least. You can submit it in another PR if you want, but note that it probably isn't something we want to support.)

projectgus reviewed (commented):
Thanks @Vincent1-python for working further on this support, I see it's mostly passing CI so I've given it a quick look. The support is impressive, but it still needs significant work before we can consider merging it.

The major tasks, as I understand after a quick review:

1. Remove the TinyUSB HID feature from this PR, it's not related to ESP32-P4 support.
2. Remove the FireBeetle board support and submit that in a second PR after this one, to keep this PR smaller and cleaner.
3. Translate the Chinese comments and variable names.

After that the PR will be smaller, and can be reviewed again. Feel free to ping me once you've done these things.

One other note: Updating to use the newer ESP-IDF driver APIs is very welcome, thanks - however that's a big change by itself so will need a lot of testing on the existing ESP32 chips.

Vincent1-python commented on ports/esp32/boards/ESP32_GENERIC_P4/mpconfigboard.h:
```diff
@@ -0,0 +1,46 @@
+#define MICROPY_PY_TINYUSB                      (0)
+#define MICROPY_PY_ESPNOW         	(0)
+
+#define MICROPY_PY_MACHINE_ADC              (0)
```
Yes, but I forgot to modify it

Vincent1-python commented on ports/esp32/boards/ESP32_GENERIC_P4/mpconfigboard.h:
```diff
@@ -0,0 +1,46 @@
+
+#define MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE 	(0)
+#define MICROPY_HW_ENABLE_USBDEV				(0)
+#define MICROPY_PY_TINYUSB                      (0)
```
Yes, but I forgot to modify it

kdschlosser commented on ports/esp32/machine_i2c.c:
```diff
@@ -28,19 +28,33 @@
 
-#if SOC_I2C_SUPPORT_XTAL
-#if CONFIG_XTAL_FREQ > 0
-#define I2C_SCLK_FREQ (CONFIG_XTAL_FREQ * 1000000)
```
I do not believe that I2c needs to be modified in the way it has been in order for it to work. For some reason I don't believe any change to the I2C needs to be made.

projectgus commented on ports/esp32/boards/ESP32_GENERIC_P4_WIFI/board.json:
```diff
@@ -0,0 +1,24 @@
+{
```
(So I suggest: delete the board `ESP32_GENERIC_P4_WIFI`, enable wifi on the default `ESP32_GENERIC_P4`. Then we can split into two variants later, no in this PR.)

kdschlosser replied:
I have a P4 sitting right here that has no WiFi.

projectgus replied:
> I have a P4 sitting right here that has no WiFi.

Good, I guess? My comment above clearly says that we want to have `wifi` and `nowifi` variants for the ESP32_GENERIC_P4, I'm just saying that we can merge this PR with a single variant - to keep it simple - and split into two variants in a follow-up PR later.

It's also an easier task for someone familiar with MicroPython and ESP-IDF (like me, or you if you're feeling generous) to make that change in a follow-up.

@Vincent1-python The latest version of the generic board looks good to me, but please name it `ESP32_GENERIC_P4`.

projectgus commented on ports/esp32/boards/ESP32_GENERIC_P4/sdkconfig.board:
```diff
@@ -0,0 +1,58 @@
+# Only on: ESP32, ESP32S2, ESP32S3
+#CONFIG_BT_NIMBLE_PINNED_TO_CORE_0=n
+#CONFIG_BT_NIMBLE_PINNED_TO_CORE_1=y
+#CONFIG_BT_NIMBLE_PINNED_TO_CORE=1
```
You've clicked Resolve again here without making the requested change or giving a reason for keeping it. Please don't do that.

Vincent1-python commented:
@projectgus  I soon found that if I delete the LDO driven by SD card, there will be many development boards that can't use SD card. I think I should add it back, which is not unique to my development board.

kdschlosser commented:
I think that we need to just get past the core of adding the P4. Anything that is additional like the LDO should be done in another PR just in case it causes some kind of a goof. Only make changes that are 100 needed to get the P4 working. Then in separate PR's add in the extra bits and pieces where things that could effect other ESP32 boards and/or other MCUs afterwards.

I know that the P4 was released some time ago there really hasn't been any available until recently for boards to be developed. It's going to be a bit longer until we see a more widespread use of the P4. Plus there was just a release of MicroPython so it's going to be a while until the next release is done so there is ample time to get some feedback on what should be added.

Just concentrate on the core things to get it to compile and be able to run and then we can do PR's for each one of the additional pieces.

As far as the LDO is concerned. That is going to be a tricky one because we want to be able to add the feature without changing the API at all. Maybe adding an extra method for the P4 to be able to turn on and off the LDO as needed.

projectgus commented:
@Vincent1-python I see you pulled all the `machine_i2c.c` changes out of this PR, but didn't submit them anywhere else. Did you find that the ESP32-P4 I2C works correctly with the existing code?

Vincent1-python commented:
> @Vincent1-python I see you pulled all the `machine_i2c.c` changes out of this PR, but didn't submit them anywhere else. Did you find that the ESP32-P4 I2C works correctly with the existing code?

There is nothing wrong with the test so far.

kdschlosser commented:
the I2C does need to be upgraded and I do suggest that @Vincent1-python submit a PR for those changes since it would bring the I2C code up to the current API in the ESP-IDF.

kdschlosser commented:
The LDO thing I think the best approach is to add in a new set of classes into the esp32 module.

Vincent1-python commented:
> the I2C does need to be upgraded and I do suggest that @Vincent1-python submit a PR for those changes since it would bring the I2C code up to the current API in the ESP-IDF.

OK.

kdschlosser commented:
the only suggestion I have to make is this...

in `ports/esp32/boards/ESP32_GENERIC_P4` add a file named `mpconfigvariant_C6_WIFI.cmake`

In that file add the following code....

```cmake
set(IDF_TARGET esp32p4)

set(SDKCONFIG_DEFAULTS
    boards/sdkconfig.base
    boards/sdkconfig.p4
    boards/sdkconfig.p4usb
    boards/ESP32_GENERIC_P4/sdkconfig.board
    boards/ESP32_GENERIC_P4/sdkconfig.c6_wifi
)
```
In that same folder rename `sdkconfig.board` to `sdkconfig.c_wifi` and then make a new `sdkconfig.board` file and leave it empty. Doing that will add `BOARD_VARIANT=C6_WIFI` as a build command. 

I don't remember what C series that is used as the wireless module so adjust as needed so the correct SOC is used. If more than one SOC is able to support the function of being a wireless module for the P4 then I would repeat this process for each of the SOC's that can be used and make any alterations to the `sdkconfig` file specific to that variant if changes need to be made. If they all use the same settings then remove the `c6_` from the extension for the `sdkconfig` and point all of the `mpconfigvariant_*_WIFI.cmake` files to that common WiFi `sdkconfig` file. 

IDK if this is something that @projectgus will agree on or not.

we can also flip flop how it works so as a default an external wireless module is enabled and add a board variant that disables it.

I just did a check and it appears that you would need to add a variant to turn it on and not off. You also have to create a variant explicitly for each of the following MCU's to enable it for and you would have to set a macro that is specific to the MCU being used..

Here are the config settings that are specific to the external MCU being used. These config flags need to be set according to what is being used as an external module. IDK if there is a way to automatically detect the module that is being used. I would imaging the external module and the P4 would chare the same USB so both devices should appear when the board gets plugged into a computer. If that is the case then that is able to be used to determine the external module and the build can be made to automatically detect it. However it would require that the MCU be plugged into the compiling machine for it to work.

Gonna have to make a board variant for each of the MCUs seen below and also an `sdkconfig` for each of the board variants to set the proper config flag in. 
```
CONFIG_SLAVE_IDF_TARGET_ESP32
CONFIG_SLAVE_IDF_TARGET_ESP32C2
CONFIG_SLAVE_IDF_TARGET_ESP32C3
CONFIG_SLAVE_IDF_TARGET_ESP32C5
CONFIG_SLAVE_IDF_TARGET_ESP32C6
CONFIG_SLAVE_IDF_TARGET_ESP32S2
CONFIG_SLAVE_IDF_TARGET_ESP32S3
CONFIG_SLAVE_IDF_TARGET_ESP32H2
CONFIG_SLAVE_IDF_TARGET_ESP32P4
CONFIG_SLAVE_IDF_TARGET_ESP32C61
```

kdschlosser commented:
From what is said for the LDO seen here...

https://docs.espressif.com/projects/esp-idf/en/stable/esp32p4/api-reference/peripherals/ldo_regulator.html

the LDO is not something that is locked to only an SDCard reader. It is simply 4 outputs that the user is able to programmatically control. The LDO's are able to be used for anything that the user may want to turn on and off. The LDO falls more in line with being a high current GPIO pin that can only be used for output. 

It looks like there are only 2 channels available for user use, 3 & 4 and those channels have a max current rating of 50mA.

Vincent1-python commented:
> CONFIG_SLAVE_IDF_TARGET_ESP32C2
> CONFIG_SLAVE_IDF_TARGET_ESP32C3
> CONFIG_SLAVE_IDF_TARGET_ESP32C5
> CONFIG_SLAVE_IDF_TARGET_ESP32C6

It seems that there are only ESP32C2, ESP32C5 and ESP32C6.

Vincent1-python commented:
> > CONFIG_SLAVE_IDF_TARGET_ESP32C2
> > CONFIG_SLAVE_IDF_TARGET_ESP32C3
> > CONFIG_SLAVE_IDF_TARGET_ESP32C5
> > CONFIG_SLAVE_IDF_TARGET_ESP32C6
> 
> It seems that there are only ESP32C2, ESP32C5 and ESP32C6.

	choice ESP_HOSTED_P4_DEV_BOARD
		bool "Configure GPIOs for Development Board"
		depends on IDF_TARGET_ESP32P4
		default ESP_HOSTED_P4_DEV_BOARD_NONE
		help
			Sets GPIOs used for transport on these ESP32-P4 Development Boards
			- ESP32-P4 Function_EV_Board with on-board ESP32-C6
			- ESP32-P4 Function_EV_Board with ESP32-C5 on a EV board
			- ESP32-P4 Function_EV_Board with ESP32-C2 on a EV board
			- ESP32-P4-C5 Code Board with on-board ESP32-C5
			Select "No development board" if you want to modify GPIOs used

		config ESP_HOSTED_P4_DEV_BOARD_NONE
			bool "No development board"

		config ESP_HOSTED_P4_DEV_BOARD_FUNC_BOARD
			depends on SLAVE_IDF_TARGET_ESP32C6 || SLAVE_IDF_TARGET_ESP32C5 || SLAVE_IDF_TARGET_ESP32C2
			bool "ESP32-P4-Function-EV-Board with on-board C6 or C5/C2 on EV board"

		config ESP_HOSTED_P4_C5_CORE_BOARD
			depends on SLAVE_IDF_TARGET_ESP32C5
			bool "ESP32-P4-C5 Core Board with on-board C5"

kdschlosser commented:
> No, it needs to be initialized and transferred to the SD card drive.

But this would be better done if what is passed to the SPI constructor is a type of Pin class for the LDO.

Vincent1-python commented:
> OK you are correct about the LDO with the SDCard. The driver needs to be able to shift the voltage between 1.8 and 3.3 volts in order to support UHS-I which is kind of funny because in the ESP-IDF docs it states that the ESP-IDF doesn't currently support UHS-I
> 
> https://docs.espressif.com/projects/esp-idf/en/release-v5.4/esp32p4/api-reference/peripherals/sdmmc_host.html#overview
> 
> ```
> 
> 
>     SDMMC_HOST_SLOT_1 is routed via GPIO Matrix. This means that any GPIO may be used for each of the SD card signals. It is for non UHS-I usage.
> 
>     SDMMC_HOST_SLOT_0 is dedicated to UHS-I mode, which is not yet supported in the driver.
> ```
> 
> last line there. It is not yet supported. So yes you can simply latch that LCD to 3.3v and the SD Card reader is going to function without any issues. UHS-II cards are able to work at 3.3 volts without needing the 1.8v.
> 
> So basically there is no point to using the LDO for this purpose because the driver doesn't support UHS-I. Adding it doesn't appear like it needs to be done at this point in time.

I hope someone else will come to the PR function.

Vincent1-python commented:
So now my PR can be merged? Still need to be revised. @projectgus @kdschlosser

kdschlosser commented:
my revisions don't mean a hill of beans in terms of being able to merge the PR. Mine is just for informational purposes I guess.

@projectgus needs to chime in on the whole LDO thing and how it should be handled because that is going to be something that causes problems with the P4 if the LDO is used to control the IO voltage for the SDReader. I think the best thing is going to be adding a macro for what LDO channel is to be used and if the macro has been defined then the LDO code is added and if it has not been defined then it's not added.

ftab commented:
> if the LDO is used to control the IO voltage for the SDReader

It's going to depend on the board. On the p4, GPIO39-48 draw their power from VDD_IO_5, and on the p4 function ev board, that pin gets its power from LDO_VO4, so programmable regulator number 4 needs to be turned on in order for the SD card (hooked up to GPIO 39-44 on this board) to function.

projectgus commented:
> > if the LDO is used to control the IO voltage for the SDReader
> 
> It's going to depend on the board. On the p4, GPIO39-48 draw their power from VDD_IO_5, and on the p4 function ev board, that pin gets its power from LDO_VO4, so programmable regulator number 4 needs to be turned on in order for the SD card (hooked up to GPIO 39-44 on this board) to function.

Exactly! There are already other boards which don't use VO4 for `VDD_IO_5` (see [my comment here](https://github.com/micropython/micropython/pull/17951#issuecomment-3208588767). Also some boards may require VO4 to be 1.8V or 2.5V so we can't safely turn it on to 3.3V on every board.

My suggestion is:

1) When you add FireBeetle 2 support, add a board_init function that turns the LDO on to 3.3V on boot. This will allow IO48 and the SD card to work on this board.
2) We should also add a machine module function to set the LDO on/off or change the voltage from Python code, that can be used on any board. Can do this later, though.
3) For UHS-I support, add a parameter to the SDCard constructor so the SD card driver can set the voltage. Again, this can be done later.

projectgus reviewed (commented):
This generally looks good to me, thanks @Vincent1-python and @kdschlosser. I like the board variants for the different Wi-Fi chips.

The path to merging looks to me like:

1) Merge I2C driver changes in #17971 and rebase this PR.
2) I want to do some testing of my own (I have two P4 boards here now, the M5Stack Tab5 and a DFRobot Firebeetle 2). I should have time to do that next week.
3) We currently make our CI and nightly builds with ESP-IDF V5.4.2. We could merge this PR and add builds for P4 later, or we can update all of our esp32 builds to ESP-IDF V5.5 first. Not sure about this one, but this is for me and Damien to worry about.

kdschlosser commented:
You need to merge the current master into this PR. when you made this PR you should have made a branch for the PR and then made your changes to that branch. This way you can pull any changes from the master branch that is upstream and merge those changes with your master branch. then you can merge the changes from your master branch into the PR branch. that synchronizes everything. 

If you do not know how to do this then I would revert the last commits you made and have someone with write permission merge the master branch into this PR.

robert-hh commented:
I managed to build the firmware for ESP32P4_GENERIC, but had to disable bluetooth and nimble. Otherwise some nimble include files from micropython-lib were not found. REPL at UART, but not at USB.

agatti commented:
Has anyone tested the native emitter if it works fine on this board?  In theory it should be enabled by default but maybe relevant tests haven't been looked at :)

If the emitter works, it'd also help if this can be rebased to the latest master and if somebody could check if `tests/micropython` tests work with `MICROPY_EMIT_RV32_ZBA` enabled.

The preliminary datasheet at https://www.espressif.com/sites/default/files/documentation/esp32-p4_datasheet_en.pdf only mentions a subset of commonly available extensions on section 4.1.1.1 - maybe they're in there but not yet documented.

robert-hh commented:
No problem. My steps:

1. Fetch PR17951 and check out out.
2. Check out esp-idf v5.5.1, run the install script and update the submodules
3. Clean the ports/esp32 directory from all previous builds. 
4. Run `make BOARD=ESP32_GENERIC_P4`

When the compile start, it show a notice about the various modules:
```
NOTICE: [1/8] espressif/eppp_link (1.1.2)
NOTICE: [2/8] espressif/esp_hosted (2.2.4)
NOTICE: [3/8] espressif/esp_serial_slave_link (1.1.1~1)
NOTICE: [4/8] espressif/esp_tinyusb (1.7.6)
NOTICE: [5/8] espressif/esp_wifi_remote (0.15.2)
NOTICE: [6/8] espressif/mdns (1.1.0)
NOTICE: [7/8] espressif/tinyusb (0.18.0~4)
NOTICE: [8/8] idf (5.5.1)
```
The compile & build of the esp32p4 image seems to be OK:
```
Creating esp32p4 image...
Merged 2 ELF sections
Successfully created esp32p4 image.
```
It fails at the step:
```
[1137/1410] Generating ../../genhdr/qstr.i.last
FAILED: genhdr/qstr.i.last /home/robert/Downloads/MicroPython/micropython/ports/esp32/build-ESP32_GENERIC_P4/genhdr/qstr.i.last 
/bin/sh esp-idf/main/CMakeFiles/qstr.i.last-93c470f.sh 82878c92c2052159
/home/robert/Downloads/MicroPython/micropython/extmod/nimble/modbluetooth_nimble.c:38:10: fatal error: host/ble_hs.h: No such file or directory
   38 | #include "host/ble_hs.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.
/home/robert/Downloads/MicroPython/micropython/ports/esp32/mpnimbleport.c:35:10: fatal error: nimble/nimble_port.h: No such file or directory
   35 | #include "nimble/nimble_port.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
```
The files exist in micropython/lib/mynewt-nibmle/... If I look into the command of the step (which is about 47k long), the directoris with the two missing files are not included. The build seems to use the nimble stack from esp-idf and not the one from micropython/lib.


If I set both `MICROPY_PY_BLUETOOTH` and `MICROPY_PY_BLUETOOTH_NIMBLE` to 0 in mpconfigport.h, the firmware builds & runs. WiFi works.

robert-hh commented:
@Vincent1-python  Did you manage to get ETH working? And if yes, which LAN settings did you use? The connections of my board between the IP101 PHY and the MCU  are the same as ones used by the "official" epressif DEV board. But still the device cannot be registered.

robert-hh commented:
To get ETH working only small changes to network_lan.c are required by just adding ` || CONFIG_IDF_TARGET_ESP32P4` to 4 lines of the code. The diff is below. The performance is solid.  The command to set up the LAN interface is e.g.:

```
lan=network.LAN(mdc=machine.Pin(31), mdio=machine.Pin(52),
    phy_type=network.PHY_IP101, phy_addr=1, reset=machine.Pin(51),
    ref_clk_mode=machine.Pin.IN, ref_clk=machine.Pin(50))
```

```
diff --git a/ports/esp32/network_lan.c b/ports/esp32/network_lan.c
index 309ee0b14..2e3743e39 100644
--- a/ports/esp32/network_lan.c
+++ b/ports/esp32/network_lan.c
@@ -182,13 +182,13 @@ static mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
     }
 
     eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
-    #if CONFIG_IDF_TARGET_ESP32
+    #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32P4
     eth_esp32_emac_config_t esp32_config = ETH_ESP32_EMAC_DEFAULT_CONFIG();
     #endif
 
     esp_eth_mac_t *mac = NULL;
 
-    #if CONFIG_IDF_TARGET_ESP32
+    #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32P4
     // Dynamic ref_clk configuration.
     if (args[ARG_ref_clk_mode].u_int != -1) {
         // Map the GPIO_MODE constants to EMAC_CLK constants.
@@ -223,7 +223,7 @@ static mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
     #endif
 
     switch (args[ARG_phy_type].u_int) {
-        #if CONFIG_IDF_TARGET_ESP32
+        #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32P4
         case PHY_LAN8710:
         case PHY_LAN8720:
             self->phy = esp_eth_phy_new_lan87xx(&phy_config);
@@ -286,7 +286,7 @@ static mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
         #endif
     }
 
-    #if CONFIG_IDF_TARGET_ESP32
+    #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32P4
     if (!IS_SPI_PHY(args[ARG_phy_type].u_int)) {
         if (self->mdc_pin == -1 || self->mdio_pin == -1) {
             mp_raise_ValueError(MP_ERROR_TEXT("mdc and mdio must be specified"));
```

Vincent1-python commented:
？What's going on? CI doesn't install the components that come with ESP-IDF?

robert-hh commented:
I pulled the last state of your PR. Still I have to disable bluetooth and change machine_uart.c to get it build. Not sure what's different in your build set-up. 
One small remark besides this:
For other ESP32 models the dependencies.lock file is in the lockfiles directory. That's where it is expected by CMakeLists.txt. So I moved dependencies.lock to lockfiles/dependencies.lock.esp32p4.

Vincent1-python commented:
So now I need to wait for the ci of esp32? Or will you help me retest after I get better?

Message ID: ***@***.***>

robert-hh commented:
> It seems that this cannot be because the corresponding wifi chips of these three files are different.

Is then the API different, or are these differences addressed by the esp_hosted firmware in the C2/C5/C6 co-processors, such that the WiFi API for the P4 is the same?

> So now I need to wait for the ci of esp32? Or will you help me retest after I get better?

We have to wait for final integration until MicroPython supports edp-idf  v5.5.1 or later. Besides that, I can help with testing.

Vincent1-python commented:
The settings of this chip are different only from the settings of the chip. Everything else is the same, and the api has not changed.Message ID: ***@***.***>

dpgeorge commented:
The esp32 has now moved to IDF v5.5.1.

@Vincent1-python can you please rebase this PR on the latest master?

Vincent1-python commented:
It's alright now

projectgus commented:
@Vincent1-python I'm not able to build your latest version of this branch (commit c01de20) to test it. I get one error trying to build `make BOARD=ESP32_GENERIC_P4` board and a different error if I try to build `make BOARD=ESP32_GENERIC_P4 BOARD_VARIANT=C6_WIFI`.

Does this branch build for you after you updated it?

projectgus committed:
esp32/usb_serial_jtag: Flush usb_serial_jtag TXFIFO from ISR.

This was necessary to un-wedge the USJ TX path on ESP32-P4, I think because
the bootloader prints a lot on this chip. I think it might be possible to
hit it on other chips, though.

The implementation is based on the ESP-IDF driver, which will always add an
extra flush when the TXFIFO is empty in case the host is expecting a ZLP.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
shared/tinyusb: Remove USBD_RHPORT constant.

TinyUSB defines TUD_OPT_RHPORT which is the same thing, make
shorter definition RHPORT in the two files which use it.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
esp32: Support building with network and/or bluetooth disabled.

(and a smaller binary size as a result)

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus reviewed (commented):
Thanks for keeping this PR up to date and addressing reviews, @Vincent1-python.

I've just done a rebase and squash, including updating the USB config to work with #18407. Also:

* Cleaned up the default variant so it builds, and reduced the binary size by excluding modules for network, socket, bluetooth, etc.
* Fixed an issue where USB Serial/JTAG output could get wedged if host was expecting a ZLP. I think this is unrelated to ESP32-P4, I think just luck that this chip prints a lot from the bootloader and happened to trigger it.
* Added CI builds for the default and the C6_WIFI variant, as they're quite different.
* Dropped the C2_WIFI variant as I can't find any board with this config. It's easy to add back later if needed.

Ran some tests here on a DFRobot P4 board (`C6_WIFI` variant):

* Both USB ports work as REPLs (TinyUSB on the highspeed port and USB Serial/JTAG on the other port).
* `multi_net` tests pass except for `udp_recv_dontwait.py` which seems like it may be broken in master.
* `multi_bluetooth` tests pass when the P4 is instance0 but some tests fail when the P4 is instance1, will follow up later.

Apart from the failing tests, there are some things it'd be good to follow up later:

* The default variant binary size is about 20% more than it needs to be, because it's still linking esp_hosted. Solving this in a maintainable way is harder than expected.
* The `test_serial.py` test is really slow in the OUT direction (host to device) on the P4, slower than most other boards - which seems wrong as the P4 is doing High Speed USB.

dpgeorge commented:
I now have a FireBeetle 2, which is an ESP32-P4 with ESP32-C6 coprocessor.

I tested with ESP32_GENERIC_P4 with the base variant and the C6_WIFI variant.  Both seem to work well running the test suite, including WLAN and BLE, and the REPL is available on the UART and native USB ports.

Things that don't work:
- running any native code (`@micropython.native`) crashes the board
- some BLE events are out of order

I think we should disable the native emitter for the P4 (or maybe it's a quick fix...).  But otherwise things look good!

agatti commented:
> running any native code (@micropython.native) crashes the board

I assume this also occurs for viper code too.

Without a board to test, the most two probable causes are more region protection bit issues (just like when support for the ESP32C6 was added), or a cache flush that's missing.  There should be no unsupported opcodes or anything unexpected otherwise.

If it's a matter of a missing cache flush, can you try to add `asm volatile ("fence\n");` in `mp_emit_glue_assign_native`?  If that works, then you may want to replace `fence` with `fence.i`, so just the instruction cache gets flushed instead of both instruction and data caches.

If that doesn't work, is it possible to get a crash/register dump of running `def f(): return 0` with the emitter?  The key information here is the values of `MSTATUS`, `MSTATUSH`, `MCAUSE`, `MEPC`, and `MTVAL`.

dpgeorge commented:
> Just 2c, there is an `esp_cache_msync` function which can help perform cache writeback

Ah, excellent!  Thanks for the hint @igrr .  I added a call to `esp_cache_msync()` and it now works!

I pushed a commit for that change to this PR.

dpgeorge committed:
py/emitglue: Add hook for RV32 arch to flush D-cache for native emitter.

Eventually this cache flushing mechanism should be generalised to work the
same way for all architectures.  But for now, this allows ESP32 RV32 SoCs
to flush the D-cache whenn needed.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
esp32/mphalport: Enable D-cache flushing on P4 for native code.

This is necessary to get native code running on the ESP32-P4.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
@exussum12 it looks like you have an early revision P4.  You'll probably need to change the ESP32P4_REV_MIN option to ESP32P4_REV_MIN_0 (it defaults to ESP32P4_REV_MIN_1).

Or, get newer hardware.

dpgeorge commented:
Thanks @igrr for the details about the revisions.  It sounds like we should support rev 0.0, so I've opened a PR for that at #18512.

My testing shows that it makes pretty much no difference to the firmware when supporting rev 0.0, which is good!

> I also want to give a heads-up that P4 revision 3.0 is not backward-compatible, meaning that it won't be possible to build a binary with support for both rev<3.0 and >=3.0. Therefore I would recommend introducing variants of binaries for rev<3 and >=3, and eventually removing rev<3 binaries when most of the boards on the market are rev>=3.

OK, thanks for the heads up!  When rev 3.0 is out we will probably make that the default variant, and then a separate variant for rev<3.
